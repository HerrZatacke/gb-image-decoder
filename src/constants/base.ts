import { RGBNPalette } from '../Types';
import { BlendMode } from './blendModes';


export const BLACK = 'FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFF';
export const WHITE = '00000000000000000000000000000000';

export const SKIP_LINE = 'skip';

export const TILE_PIXEL_WIDTH = 8;

export const TILE_PIXEL_HEIGHT = 8;

export const TILES_PER_LINE = 20;

export const TILES_PER_COLUMN = 18;

export const FRAME_WIDTH = 2;

export const BW_PALETTE: number[] = [0xffffff, 0xaaaaaa, 0x555555, 0x000000];

export const BW_PALETTE_HEX: string[] = ['#ffffff', '#aaaaaa', '#555555', '#000000'];

export const BLACK_LINE: string[] = Array(TILES_PER_LINE).fill(BLACK);

export const WHITE_LINE: string[] = Array(TILES_PER_LINE).fill(WHITE);

export const RGBN_SHADES = [0, 0x55, 0xaa, 0xff];

export const defaultPalette: RGBNPalette = {
  r: RGBN_SHADES,
  g: RGBN_SHADES,
  b: RGBN_SHADES,
  n: RGBN_SHADES,
  blend: BlendMode.MULTIPLY,
};
