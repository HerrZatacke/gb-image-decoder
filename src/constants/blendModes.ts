export enum BlendMode {
  NORMAL = 'normal', // skip N-Layer
  NORMAL_S = 'normal_s', // This is the default setting and draws new shapes on top of the existing canvas content.
  LIGHTEN = 'lighten', // Retains the lightest pixels of both layers.
  SCREEN = 'screen', // The pixels are inverted, multiplied, and inverted again. A lighter picture is the result (opposite of multiply)
  DODGE = 'dodge', // Divides the bottom layer by the inverted top layer.
  ADDITION = 'addition', // Where both shapes overlap, the color is determined by adding color values.
  DARKEN = 'darken', // Retains the darkest pixels of both layers.
  MULTIPLY = 'multiply', // The pixels of the top layer are multiplied with the corresponding pixels of the bottom layer. A darker picture is the result.
  BURN = 'burn', // Divides the inverted bottom layer by the top layer, and then inverts the result.
  OVERLAY = 'overlay', // A combination of multiply and screen. Dark parts on the base layer become darker, and light parts become lighter.
  SOFTLIGHT = 'softlight', // A softer version of hard-light. Pure black or white does not result in pure black or white.
  HARDLIGHT = 'hardlight', // Like overlay, a combination of multiply and screen — but instead with the top layer and bottom layer swapped.
  DIFFERENCE = 'difference', // Subtracts the bottom layer from the top layer — or the other way round — to always get a positive value.
  EXCLUSION = 'exclusion', // Like difference, but with lower contrast.
}

export const blendModeNewName = (oldName?: BlendMode): GlobalCompositeOperation => {
  switch (oldName) {
    case BlendMode.NORMAL:
      return 'source-over';
    case BlendMode.NORMAL_S:
      return 'source-over';
    case BlendMode.LIGHTEN:
      return 'lighten';
    case BlendMode.SCREEN:
      return 'screen';
    case BlendMode.DODGE:
      return 'color-dodge';
    case BlendMode.ADDITION:
      return 'lighter';
    case BlendMode.DARKEN:
      return 'darken';
    case BlendMode.MULTIPLY:
      return 'multiply';
    case BlendMode.BURN:
      return 'color-burn';
    case BlendMode.OVERLAY:
      return 'overlay';
    case BlendMode.SOFTLIGHT:
      return 'soft-light';
    case BlendMode.HARDLIGHT:
      return 'hard-light';
    case BlendMode.DIFFERENCE:
      return 'difference';
    case BlendMode.EXCLUSION:
      return 'exclusion';
    default:
      return 'multiply';
  }
};
