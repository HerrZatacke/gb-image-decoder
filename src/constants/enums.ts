export enum ChannelKey {
  R = 'r',
  G = 'g',
  B = 'b',
  N = 'n',
}

export enum ExportFrameMode {
  FRAMEMODE_KEEP = 'keep',
  FRAMEMODE_CROP = 'crop',
  FRAMEMODE_SQUARE_BLACK = 'square_black',
  FRAMEMODE_SQUARE_WHITE = 'square_white',
}

export enum Rotation {
  DEG_0 = 0,
  DEG_90 = 1,
  DEG_180 = 2,
  DEG_270 = 3,
}

export const channels = [ChannelKey.R, ChannelKey.G, ChannelKey.B, ChannelKey.N];
