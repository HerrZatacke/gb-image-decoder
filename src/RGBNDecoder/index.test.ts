/* eslint-disable max-len */
import { beforeEach, describe, expect, test } from 'vitest';
import { hash } from 'ohash';
import { BlendMode, DecoderOptions, ExportFrameMode, RGBNTiles } from '../index';
import { RGBNDecoder } from './index';
import tiles1x1 from '../../test/data/tiles/rgbn/1x1';
import tiles16x14 from '../../test/data/tiles/rgbn/16x14';
import tiles20x18 from '../../test/data/tiles/rgbn/20x18';
import tiles20x86 from '../../test/data/tiles/rgbn/20x86';
import { rgbMultiply } from '../../test/data/palettes';
import { TILES_PER_LINE } from '../constants/base';
import { writeImageFileFromImageData } from '../../test/helpers/writeImageFileFromImageData';
import { getCanvasImageData } from '../../test/helpers/ getCanvasImageData';


const TEST_SCALE_FACTOR = 2;

const tileSets: Record<string, RGBNTiles> = {
  '1x1': tiles1x1,
  '16x14': tiles16x14,
  '20x18': tiles20x18,
  '20x86': tiles20x86,
};

describe('RGBN Decoder', () => {
  let defaultDecoderOptions: DecoderOptions = {};
  let canvas: HTMLCanvasElement;

  beforeEach(() => {
    canvas = document.createElement('canvas');
    defaultDecoderOptions = {
      tilesPerLine: TILES_PER_LINE,
    };
  });

  describe.each([
    { tileKey: '1x1', tilesPerLine: 1 },
    { tileKey: '16x14', tilesPerLine: 16 },
    { tileKey: '20x18', tilesPerLine: 20 },
    { tileKey: '20x86', tilesPerLine: 20 },
  ])('using $tileKey raw data', async ({ tileKey, tilesPerLine }) => {
    const constructDecoderOptions = {
      ...defaultDecoderOptions,
      tilesPerLine,
    };

    test.each([
      { blendMode: BlendMode.NORMAL },
      { blendMode: BlendMode.NORMAL_S },
      { blendMode: BlendMode.LIGHTEN },
      { blendMode: BlendMode.SCREEN },
      { blendMode: BlendMode.DODGE },
      { blendMode: BlendMode.ADDITION },
      { blendMode: BlendMode.DARKEN },
      { blendMode: BlendMode.MULTIPLY },
      { blendMode: BlendMode.BURN },
      { blendMode: BlendMode.OVERLAY },
      { blendMode: BlendMode.SOFTLIGHT },
      { blendMode: BlendMode.HARDLIGHT },
      { blendMode: BlendMode.DIFFERENCE },
      { blendMode: BlendMode.EXCLUSION },
    ])('renders on to a given canvas with blendmode: $blendMode', async ({ blendMode }) => {
      const decoder = new RGBNDecoder(constructDecoderOptions);

      decoder.update({
        canvas,
        palette: {
          ...rgbMultiply,
          blend: blendMode,
        },
        tiles: tileSets[tileKey],
      });

      await writeImageFileFromImageData(`rgb_${tileKey}_default_${blendMode}.png`, canvas);

      expect(hash(getCanvasImageData(canvas).data)).toMatchSnapshot();
    });

    test.each([
      { frameMode: ExportFrameMode.FRAMEMODE_KEEP },
      { frameMode: ExportFrameMode.FRAMEMODE_CROP },
      { frameMode: ExportFrameMode.FRAMEMODE_SQUARE_BLACK },
      { frameMode: ExportFrameMode.FRAMEMODE_SQUARE_WHITE },
    ])('renders a scaled canvas with framemode: $frameMode', async ({ frameMode }) => {
      const decoder = new RGBNDecoder(constructDecoderOptions);

      decoder.update({
        canvas: null,
        palette: rgbMultiply,
        tiles: tileSets[tileKey],
      });

      const scaledCanvas = decoder.getScaledCanvas(TEST_SCALE_FACTOR, frameMode);

      await writeImageFileFromImageData(`rgb_${tileKey}_scaled-${frameMode}.png`, scaledCanvas);

      expect(hash(getCanvasImageData(scaledCanvas).data)).toMatchSnapshot();
    });
  });

});
