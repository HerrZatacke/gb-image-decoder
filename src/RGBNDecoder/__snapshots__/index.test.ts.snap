// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`RGBN Decoder > using '1x1' raw data > renders a scaled canvas with framemode: 'crop' 1`] = `"tl22VltTtX"`;

exports[`RGBN Decoder > using '1x1' raw data > renders a scaled canvas with framemode: 'keep' 1`] = `"tl22VltTtX"`;

exports[`RGBN Decoder > using '1x1' raw data > renders a scaled canvas with framemode: 'square_black' 1`] = `"tl22VltTtX"`;

exports[`RGBN Decoder > using '1x1' raw data > renders a scaled canvas with framemode: 'square_white' 1`] = `"tl22VltTtX"`;

exports[`RGBN Decoder > using '1x1' raw data > renders on to a given canvas with blendmode: 'addition' 1`] = `"dD1mCIk0N0"`;

exports[`RGBN Decoder > using '1x1' raw data > renders on to a given canvas with blendmode: 'burn' 1`] = `"dx877DjL8f"`;

exports[`RGBN Decoder > using '1x1' raw data > renders on to a given canvas with blendmode: 'darken' 1`] = `"ZyJU7P6FW4"`;

exports[`RGBN Decoder > using '1x1' raw data > renders on to a given canvas with blendmode: 'difference' 1`] = `"USg5vX1EAt"`;

exports[`RGBN Decoder > using '1x1' raw data > renders on to a given canvas with blendmode: 'dodge' 1`] = `"NH46xfDmCh"`;

exports[`RGBN Decoder > using '1x1' raw data > renders on to a given canvas with blendmode: 'exclusion' 1`] = `"BUcIbSaN65"`;

exports[`RGBN Decoder > using '1x1' raw data > renders on to a given canvas with blendmode: 'hardlight' 1`] = `"blcDIFp5Q5"`;

exports[`RGBN Decoder > using '1x1' raw data > renders on to a given canvas with blendmode: 'lighten' 1`] = `"paYkGoRxwf"`;

exports[`RGBN Decoder > using '1x1' raw data > renders on to a given canvas with blendmode: 'multiply' 1`] = `"ozHLzBUmuH"`;

exports[`RGBN Decoder > using '1x1' raw data > renders on to a given canvas with blendmode: 'normal' 1`] = `"m5H8uEJPdf"`;

exports[`RGBN Decoder > using '1x1' raw data > renders on to a given canvas with blendmode: 'normal_s' 1`] = `"5VnMSYtxZD"`;

exports[`RGBN Decoder > using '1x1' raw data > renders on to a given canvas with blendmode: 'overlay' 1`] = `"BDxtcMEdyQ"`;

exports[`RGBN Decoder > using '1x1' raw data > renders on to a given canvas with blendmode: 'screen' 1`] = `"WhwZ6pK5sL"`;

exports[`RGBN Decoder > using '1x1' raw data > renders on to a given canvas with blendmode: 'softlight' 1`] = `"RrLuM96mdz"`;

exports[`RGBN Decoder > using '16x14' raw data > renders a scaled canvas with framemode: 'crop' 1`] = `"AsfJiiKBTn"`;

exports[`RGBN Decoder > using '16x14' raw data > renders a scaled canvas with framemode: 'keep' 1`] = `"AsfJiiKBTn"`;

exports[`RGBN Decoder > using '16x14' raw data > renders a scaled canvas with framemode: 'square_black' 1`] = `"AsfJiiKBTn"`;

exports[`RGBN Decoder > using '16x14' raw data > renders a scaled canvas with framemode: 'square_white' 1`] = `"AsfJiiKBTn"`;

exports[`RGBN Decoder > using '16x14' raw data > renders on to a given canvas with blendmode: 'addition' 1`] = `"EamZltM9dV"`;

exports[`RGBN Decoder > using '16x14' raw data > renders on to a given canvas with blendmode: 'burn' 1`] = `"EamZltM9dV"`;

exports[`RGBN Decoder > using '16x14' raw data > renders on to a given canvas with blendmode: 'darken' 1`] = `"EamZltM9dV"`;

exports[`RGBN Decoder > using '16x14' raw data > renders on to a given canvas with blendmode: 'difference' 1`] = `"EamZltM9dV"`;

exports[`RGBN Decoder > using '16x14' raw data > renders on to a given canvas with blendmode: 'dodge' 1`] = `"EamZltM9dV"`;

exports[`RGBN Decoder > using '16x14' raw data > renders on to a given canvas with blendmode: 'exclusion' 1`] = `"EamZltM9dV"`;

exports[`RGBN Decoder > using '16x14' raw data > renders on to a given canvas with blendmode: 'hardlight' 1`] = `"EamZltM9dV"`;

exports[`RGBN Decoder > using '16x14' raw data > renders on to a given canvas with blendmode: 'lighten' 1`] = `"EamZltM9dV"`;

exports[`RGBN Decoder > using '16x14' raw data > renders on to a given canvas with blendmode: 'multiply' 1`] = `"EamZltM9dV"`;

exports[`RGBN Decoder > using '16x14' raw data > renders on to a given canvas with blendmode: 'normal' 1`] = `"EamZltM9dV"`;

exports[`RGBN Decoder > using '16x14' raw data > renders on to a given canvas with blendmode: 'normal_s' 1`] = `"EamZltM9dV"`;

exports[`RGBN Decoder > using '16x14' raw data > renders on to a given canvas with blendmode: 'overlay' 1`] = `"EamZltM9dV"`;

exports[`RGBN Decoder > using '16x14' raw data > renders on to a given canvas with blendmode: 'screen' 1`] = `"EamZltM9dV"`;

exports[`RGBN Decoder > using '16x14' raw data > renders on to a given canvas with blendmode: 'softlight' 1`] = `"EamZltM9dV"`;

exports[`RGBN Decoder > using '20x18' raw data > renders a scaled canvas with framemode: 'crop' 1`] = `"hmYcT9Z6yy"`;

exports[`RGBN Decoder > using '20x18' raw data > renders a scaled canvas with framemode: 'keep' 1`] = `"9qMjmcDXMO"`;

exports[`RGBN Decoder > using '20x18' raw data > renders a scaled canvas with framemode: 'square_black' 1`] = `"dPLuLmkGCG"`;

exports[`RGBN Decoder > using '20x18' raw data > renders a scaled canvas with framemode: 'square_white' 1`] = `"ufPQUtmZA9"`;

exports[`RGBN Decoder > using '20x18' raw data > renders on to a given canvas with blendmode: 'addition' 1`] = `"SUtVOyfWgC"`;

exports[`RGBN Decoder > using '20x18' raw data > renders on to a given canvas with blendmode: 'burn' 1`] = `"3behd63etK"`;

exports[`RGBN Decoder > using '20x18' raw data > renders on to a given canvas with blendmode: 'darken' 1`] = `"3uedQi4sWk"`;

exports[`RGBN Decoder > using '20x18' raw data > renders on to a given canvas with blendmode: 'difference' 1`] = `"aMMxqVmQPk"`;

exports[`RGBN Decoder > using '20x18' raw data > renders on to a given canvas with blendmode: 'dodge' 1`] = `"spDlQ6ZvcX"`;

exports[`RGBN Decoder > using '20x18' raw data > renders on to a given canvas with blendmode: 'exclusion' 1`] = `"1SdioevR1c"`;

exports[`RGBN Decoder > using '20x18' raw data > renders on to a given canvas with blendmode: 'hardlight' 1`] = `"X5Vs1cEek0"`;

exports[`RGBN Decoder > using '20x18' raw data > renders on to a given canvas with blendmode: 'lighten' 1`] = `"lEfRbjm46b"`;

exports[`RGBN Decoder > using '20x18' raw data > renders on to a given canvas with blendmode: 'multiply' 1`] = `"O9GiUqnsXq"`;

exports[`RGBN Decoder > using '20x18' raw data > renders on to a given canvas with blendmode: 'normal' 1`] = `"DAXovceFYc"`;

exports[`RGBN Decoder > using '20x18' raw data > renders on to a given canvas with blendmode: 'normal_s' 1`] = `"knnJ7ayc1q"`;

exports[`RGBN Decoder > using '20x18' raw data > renders on to a given canvas with blendmode: 'overlay' 1`] = `"aS9TUOMnsg"`;

exports[`RGBN Decoder > using '20x18' raw data > renders on to a given canvas with blendmode: 'screen' 1`] = `"H6qs0b7AwD"`;

exports[`RGBN Decoder > using '20x18' raw data > renders on to a given canvas with blendmode: 'softlight' 1`] = `"4Mfdp8JDYF"`;

exports[`RGBN Decoder > using '20x86' raw data > renders a scaled canvas with framemode: 'crop' 1`] = `"Bw38t41P0J"`;

exports[`RGBN Decoder > using '20x86' raw data > renders a scaled canvas with framemode: 'keep' 1`] = `"Bw38t41P0J"`;

exports[`RGBN Decoder > using '20x86' raw data > renders a scaled canvas with framemode: 'square_black' 1`] = `"Bw38t41P0J"`;

exports[`RGBN Decoder > using '20x86' raw data > renders a scaled canvas with framemode: 'square_white' 1`] = `"Bw38t41P0J"`;

exports[`RGBN Decoder > using '20x86' raw data > renders on to a given canvas with blendmode: 'addition' 1`] = `"qA5WD6FKfx"`;

exports[`RGBN Decoder > using '20x86' raw data > renders on to a given canvas with blendmode: 'burn' 1`] = `"ejYcTmysEL"`;

exports[`RGBN Decoder > using '20x86' raw data > renders on to a given canvas with blendmode: 'darken' 1`] = `"yyJPPM794f"`;

exports[`RGBN Decoder > using '20x86' raw data > renders on to a given canvas with blendmode: 'difference' 1`] = `"ZTXE77ibQf"`;

exports[`RGBN Decoder > using '20x86' raw data > renders on to a given canvas with blendmode: 'dodge' 1`] = `"oo2Rd9IyMd"`;

exports[`RGBN Decoder > using '20x86' raw data > renders on to a given canvas with blendmode: 'exclusion' 1`] = `"XwxKIFuap1"`;

exports[`RGBN Decoder > using '20x86' raw data > renders on to a given canvas with blendmode: 'hardlight' 1`] = `"sWQn6Gwtwr"`;

exports[`RGBN Decoder > using '20x86' raw data > renders on to a given canvas with blendmode: 'lighten' 1`] = `"k1nK0zK58P"`;

exports[`RGBN Decoder > using '20x86' raw data > renders on to a given canvas with blendmode: 'multiply' 1`] = `"zSJw9ZmpU6"`;

exports[`RGBN Decoder > using '20x86' raw data > renders on to a given canvas with blendmode: 'normal' 1`] = `"d5FwdEAQlr"`;

exports[`RGBN Decoder > using '20x86' raw data > renders on to a given canvas with blendmode: 'normal_s' 1`] = `"Jt0WRjMQ1H"`;

exports[`RGBN Decoder > using '20x86' raw data > renders on to a given canvas with blendmode: 'overlay' 1`] = `"D0Z1lFjACp"`;

exports[`RGBN Decoder > using '20x86' raw data > renders on to a given canvas with blendmode: 'screen' 1`] = `"eX6UHq6PLY"`;

exports[`RGBN Decoder > using '20x86' raw data > renders on to a given canvas with blendmode: 'softlight' 1`] = `"6BT1vQdokw"`;
